portfolio/src/main/java/com/google/sps/servlets/ReportServlet.java

```diff
@@ -22,39 +22,39 @@
 
   DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
-  public int incrementReportCount(String id, int numRetries) {
-    Transaction transaction = datastore.beginTransaction();
-    try {
-      try {
-        Key commentKey = KeyFactory.stringToKey(id);
-        Entity commentEntity = datastore.get(commentKey);
-        long numReports = (long) commentEntity.getProperty("numReports");
-        numReports++;
-        commentEntity.setProperty("numReports", numReports);
-        datastore.put(transaction, commentEntity);
-        transaction.commit();
-      } catch (EntityNotFoundException e) {
-        System.err.println("Key is not an entity in the database");
-      }
-      return 0;
-    } catch (ConcurrentModificationException e) {
-      if(numRetries == 0) {
-        throw e;
-      }
-      return --numRetries;
-    } finally {
-      if(transaction.isActive()) {
-        transaction.rollback();
-      }
+  public boolean incrementReportCount(String id, Transaction transaction) throws EntityNotFoundException, ConcurrentModificationException {
+    Key commentKey = KeyFactory.stringToKey(id);
+    Entity commentEntity = datastore.get(commentKey);
+    if(commentEntity == null) {
```

Comment:
get doesn't return null according to the doc. It would throw an exception. No need to do this check or have this method return a boolean since its not being used in the caller.
https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/DatastoreService.html#get-com.google.appengine.api.datastore.Key-

Also, the transaction is only grouping the put. You want to group both the get and write. Use get(transaction, commentKey) instead.
https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/DatastoreService.html#get-com.google.appengine.api.datastore.Transaction-com.google.appengine.api.datastore.Key-